compiler/compile/OMRCompilation.cpp

```diff
@@ -2079,6 +2082,53 @@ void OMR::Compilation::verifyCFG(TR::ResolvedMethodSymbol *methodSymbol)
       }
    }
 
+
+/*
+ * \brief
+ *    Check to make sure the rdbars are anchored and anchor a rdbar if it's found unanchored
+ *
+ * \note
+ *    Ideally all optimizations should anchor a rdbar when it's created.
+ *    This API is for fixing the unanchored rdbars in case some optimizations forget to
+ *    anchor the node by mistake. Optimizations should still try to generate correct rdbar
+ *    trees instead of relying on this API.
+ */
+void OMR::Compilation::verifyAndFixRdbarAnchors()
+   {
+   TR::NodeChecklist anchoredRdbarNodes(self());
+   NodeTreeTopPairList unanchoredRdbarNodes(self()->trMemory()->currentStackRegion());
+   for (TR::PreorderNodeIterator iter(self()->getStartTree(), self()); iter.currentTree() != NULL; ++iter)
+      {
+      TR::Node *node = iter.currentNode();
+      if (node->getOpCodeValue() == TR::treetop ||
+          node->getOpCode().isResolveOrNullCheck() ||
+          node->getOpCode().isStoreReg() ||
+          node->getOpCode().isAnchor())
+         {
+         if (node->getFirstChild()->getOpCode().isReadBar())
+            anchoredRdbarNodes.add(node->getFirstChild());
+         }
+      else if (node->getOpCode().isReadBar())
+         {
+         if (!anchoredRdbarNodes.contains(node))
+            {
+            TR_ASSERT(0, "node (n%dn) %p is rdbar but not anchored\n", node->getGlobalIndex(), node);
+            traceMsg(self(), "node (n%dn) %p is rdbar but not anchored\n", node->getGlobalIndex(), node);
+            unanchoredRdbarNodes.push_back(std::make_pair(node, iter.currentTree()));
+            }
+         }
+      }
+
+   for (auto it = unanchoredRdbarNodes.begin(); it != unanchoredRdbarNodes.end(); it++)
```

Comment:
do we have to save them all up or could we just do it as we go? it would take less resources to do it as we go - eg put the anchors in before the current walk treetop so as to not mess up the iterator?

Reply:
Changed to fix the treetop as we go instead of in another pass.